org.eclipse.jdt.core.compiler.batch/src/org/eclipse/jdt/internal/compiler/codegen/BranchLabel.java

```diff
@@ -145,13 +168,16 @@ protected void trackStackDepth(boolean branch) {
 			if (this.targetStackDepth != this.codeStream.stackDepth) {
 				this.codeStream.classFile.referenceBinding.scope.problemReporter().operandStackSizeInappropriate(
 						this.codeStream.classFile.referenceBinding.scope.referenceContext);
-				if (this.targetStackDepth < this.codeStream.stackDepth)
+				if (this.targetStackDepth < this.codeStream.stackDepth) {
 					this.targetStackDepth = this.codeStream.stackDepth; // FWIW, pick the higher water mark.
+					this.operandStack = this.codeStream.operandStack.copy();
+				}
 			}
-			// TODO: check that contents slot count matches targetStackDepth
+			// TODO: check that operand stacks match.
+		} else {
+			this.codeStream.stackDepth = this.targetStackDepth;
+			this.codeStream.operandStack = this.operandStack.copy();
 		}
-		this.codeStream.stackDepth = this.targetStackDepth;
-		this.codeStream.operandStack = this.switchSaveTypeBindings.copy();
 	}
 }
```

Comment:
The wrapping of the pre-existing statements 
```
                        this.codeStream.stackDepth = this.targetStackDepth;
			this.codeStream.operandStack = this.operandStack.copy();
```

is a fix unrelated to the defect. As the method [validOperandStack](https://github.com/eclipse-jdt/eclipse.jdt.core/pull/2442/files) shows, it is incorrect to expect that operand stack contents are identical along branch edges. They are only required to be compatible. (think of `cond ? new String() : new Date()` - as long as a LUB exists it is ok). So where `targetStackDepth` is known **AND** `sourceDepthKnown` is true, we should not copy over the operand stack. That could replace sharper values with blunter ones and vice versa.